src/audio/google/google_rtc_audio_processing.c

```diff
@@ -542,221 +613,239 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
+	GoogleRtcAudioProcessingProcessCapture_float32(cd->state,
+						       (const float **)cd->raw_mic_buffers,
+						       cd->refout_buffers);
+	cd->buffered_frames = 0;
+}
+
+static bool ref_stream_active(struct google_rtc_audio_processing_comp_data *cd)
+{
+	return cd->ref_comp_buffer->source &&
+		cd->ref_comp_buffer->source->state == COMP_STATE_ACTIVE;
+}
+
+static int mod_process(struct processing_module *mod, struct sof_source **sources,
+		       int num_of_sources, struct sof_sink **sinks, int num_of_sinks)
 {
 	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
-	int16_t *src, *dst, *ref;
-	uint32_t num_aec_reference_frames;
-	uint32_t num_aec_reference_bytes;
-	int num_samples_remaining;
-	int num_frames_remaining;
-	int channel;
-	int frames;
-	int nmax;
-	int ret;
-	int i, j, n;
 
-	struct input_stream_buffer *ref_streamb, *mic_streamb;
-	struct output_stream_buffer *out_streamb;
-	struct audio_stream *ref_stream, *mic_stream, *out_stream;
+	if (cd->reconfigure)
+		google_rtc_audio_processing_reconfigure(mod);
 
-	if (cd->reconfigure) {
-		ret = google_rtc_audio_processing_reconfigure(mod);
-		if (ret)
-			return ret;
-	}
+	struct sof_source *mic = sources[cd->raw_microphone_source];
+	struct sof_source *ref = sources[cd->aec_reference_source];
+	struct sof_sink *out = sinks[0];
+	bool ref_ok = ref_stream_active(cd);
 
-	ref_streamb = &input_buffers[cd->aec_reference_source];
-	ref_stream = ref_streamb->data;
-	ref = audio_stream_get_rptr(ref_stream);
-
-	num_aec_reference_frames = input_buffers[cd->aec_reference_source].size;
-	num_aec_reference_bytes = audio_stream_frame_bytes(ref_stream) * num_aec_reference_frames;
-
-	num_samples_remaining = num_aec_reference_frames * audio_stream_get_channels(ref_stream);
-	while (num_samples_remaining) {
-		nmax = audio_stream_samples_without_wrap_s16(ref_stream, ref);
-		n = MIN(num_samples_remaining, nmax);
-		for (i = 0; i < n; i += cd->num_aec_reference_channels) {
-			j = cd->num_aec_reference_channels * cd->aec_reference_frame_index;
-			for (channel = 0; channel < cd->num_aec_reference_channels; ++channel)
-				cd->aec_reference_buffer[j++] = ref[channel];
-
-			ref += audio_stream_get_channels(ref_stream);
-			++cd->aec_reference_frame_index;
-
-			if (cd->aec_reference_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
-									    cd->aec_reference_buffer);
-				cd->aec_reference_frame_index = 0;
-			}
-		}
-		num_samples_remaining -= n;
-		ref = audio_stream_wrap(ref_stream, ref);
+	/* Clear the buffer if the reference pipeline shuts off */
+	if (!ref_ok && cd->last_ref_ok)
+		bzero(arch_xtensa_cached_ptr(refoutbuf), sizeof(refoutbuf));
+
+	int fmic = source_get_data_frames_available(mic);
+	int fref = source_get_data_frames_available(ref);
+	int frames = ref_ok ? MIN(fmic, fref) : fmic;
+	int n, frames_rem;
+
+	/* If fref > fmic at the startup of the stream, we should
```

Comment:
I don't know about IPC3/LL, but for DP please don't try to process more than 10ms in a period

why: 

TICK
 - Core 0 LL finishes processing, AEC has 10ms in both buffers => ready to running
 TICK 
 - core 0 starts another LL cycle, core1 starts DP (and is late a bit- i.e. was processing IPC, does not matter why)
 - core0 finishes some processing before DP starts, there's additional 1ms data in one of inputs (or even on both)
 
Your change will eithter
 - if one input has 11ms (it may happen not only on startup!!) - additional ms will be dropped . Worse - in a following cycle the other input will have 1 extra ms ===> Glitch
 - if both have 11ms - AEC will try to proceess 11ms in single cycle. And AEC is a heavy module, it utilizes >90% of CPU core and it MUST finish in 10ms, it can easily be late ==>  glitch
 
 so - for DP don't try to process more than 10ms in a single cycle. I think even about change that won't report more data to DP module than a single period, regardless how much is in the buffer.

separate part is a desync at startup - I did not see it in my tests, when I was starting pipelines manually from shell. Please give some more details. 

Dropping data is not a good idea.

Reply:
@marcinszkudlinski could you explain why your explained possible problematic situation triggers with 10ms and not with 9 or 2 or 20?

Reply:
Agreed, not understanding this.  If we hit this code path (and we don't in practice: we did, but only because of the since-fixed buffer format issue), we have a backup.  We should at least try to clear that backup as soon as practical, otherwise latency on the capture pipeline will never recover.

Reply:
Ah, I think there's a little bit of conflation going on here.  The particular code above this comment isn't about the greedy "catch up" behavior in AEC, it's to handle the case where a playback stream is already in progress at the moment where a capture stream starts.  On mt8195, the prepare() notification arrives 20-30ms before the first data on the capture pipeline.  So the AEC component starts up with effectively a bunch of extra frames that do nothing but add latency.  That said, I don't see any other corrections in the IPC4 cases that would handle this elsewhere?  I'm pretty sure this logic is needed everywhere.  Note that it runs only once, if it discards "too much" data on the reference pipe, the net effect is just adding that much latency to the capture path (which isn't perfect, but it's tolerable and stable).

Reply:
FWIW: Per and Lionel point out that there is delay estimation built in to AEC, so this isn't really needed as long as the mismatch is bounded (and it is, by the input buffer sizes).   I removed it.